### PR TITLE
Array: non-dask-arrays and broadcasting in `ravel_multi_index`

### LINKED ISSUE #7580: dask.array.ravel_multi_index doesn't replicate numpy API
**What happened**:
`dask.array.ravel_multi_index` fails when passed a tuple of integer arrays (`AttributeError: 'tuple' object has no attribute 'map_blocks'`)

**What you expected to happen**:
The documentation (copied from the equivalent numpy function) indicates that the first input should be tuple of integer arrays, one array for each dimension.

**Minimal Complete Verifiable Example**:
```python
import numpy as np
import dask.array as dsa

arr = np.array([[3,6,6],[4,5,1]]) # Example from numpy docs
darr = dsa.from_array(arr)

np.ravel_multi_index((arr, arr), (7,7))    # This works
dsa.ravel_multi_index((darr, darr), (7,7)) # This doesn't
```

**Environment**:

- Dask version: 2021.04.0
- Python version: 3.8.8
- Operating System: Linux 4.12.14-95.57-default
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by gjoseph92**
`ravel_multi_index` assumed the input was always a dask array (#7580). This fixes that, but also resolves other things that diverged from the NumPy implementation. Now:
* Each index array can be >1D
* Index arrays are broadcast to a common shape first, so they can have different numbers of dimensions (including scalars)

With a little bit more thinking, I'm sure we could get this to use `blockwise` for better graph performance—it just wasn't working on the multidimensional/broadcasting case. But hopefully, we can rewrite `broadcast_arrays` and `stack` soon to use blockwise, so we can use these higher-level operations and get the benefits.

@Illviljan sorry, I found this issue independently and didn't see #7584 until I had already written all this... how does this seem to you?

cc @jrbourbeau @dougiesquire

Note: requires #7593 for tests to pass.

- [x] Closes #7580
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/array/tests/test_routines.py [hunk 1]
```diff
@@ -1734,17 +1746,48 @@ def test_unravel_index():
...
+    if asarray is da.from_array:
+        arr = np.asarray(arr)
+        input = da.from_array(arr, chunks=chunks)
+    else:
+        arr = input = asarray(arr)
+
     assert_eq(
-        da.ravel_multi_index(darr, **kwargs).compute(),
+        da.ravel_multi_index(input, **kwargs),
         np.ravel_multi_index(arr, **kwargs),
```
Comment (Illviljan) [inline dask/array/tests/test_routines.py:1772]:
```suggestion
        np.ravel_multi_index(arr, **kwargs),
        da.ravel_multi_index(input, **kwargs),
```
It looks better in pytest if the truth errors first, I was trying some cases where the np version also errored.

**Review by Illviljan — COMMENTED** [review]
This seems more thought out than my version so I think we'll go with this one.

File: dask/array/routines.py [hunk 2]
```diff
@@ -1747,18 +1748,35 @@ def unravel_index(indices, shape, order="C"):
...
+    if is_arraylike(multi_index):
+        index_stack = asarray(multi_index)
+    else:
+        multi_index_arrs = broadcast_arrays(*multi_index)
+        index_stack = stack(multi_index_arrs)
+
+    if len(index_stack) != len(dims):
+        raise ValueError(
+            f"parameter multi_index must be a sequence of length {len(dims)}"
+        )
```
Comment (jrbourbeau) [inline dask/array/routines.py:1769]: We might want to forgo this validation check for arrays with an unknown size

**Review by jrbourbeau — COMMENTED** [review]
Thanks @gjoseph92! One small comment, otherwise this looks great

**Comment by gjoseph92** [comment]
@jrbourbeau good point, I've now left it to downstream methods be restrictive about unknown chunksizes.

All green except for one DataFrame `test_pivot_table[sum-values1]` test on 3.8. Seems flaky, since that should be totally unrelated?

**Review by jrbourbeau — COMMENTED** [review]
Yep, agreed that looks unrelated. Just restarted CI to see if the failure is transient

**Comment by jrbourbeau** [comment]
I've restarted CI a few times and `test_pivot_table` appears to be failing consistently. I also opened a separate PR (xref https://github.com/dask/dask/pull/7629) to test CI against the current `main` branch and didn't see any test failures. That said, I don't think the changes here are related to `test_pivot_table` at all. Could I ask you investigate a bit further before we merge this?

**Comment by gjoseph92** [comment]
Yup, I'll dig into it more. This is very strange.

**Review by jrbourbeau — APPROVED** [review]
Just merged `main` into this PR and `test_pivot_table` is now passing ✨. I'm going to merge this and keep an eye out for any `test_pivot_table` failures. Thanks @gjoseph92!

**Pull request merged**